core/Enclave/enclave_hsm.cpp

```diff
@@ -489,30 +655,109 @@ sgx_status_t enclave_rsa_decrypt(const uint8_t *cmk_blob, size_t cmk_blob_size,
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+            break;
+        case EH_RSA_3072:
+            sealed_rsa_len = sgx_calc_gcm_data_size(0, 3060);
+
+            if (ciphertext == NULL || ciphertext_len < RSA_OAEP_3072_CIPHER_LENGTH) {
+                printf("ecall rsa_decrypt ciphertext len is too small.\n");
+                return SGX_ERROR_INVALID_PARAMETER;
+            }
+            break;
     }
 
-    if (plaintext_len == 0)
-    {
-        unsigned char* to;
-        RSA_private_decrypt(ciphertext_len, ciphertext, to, rsa, RSA_PKCS1_OAEP_PADDING);
-        *req_plaintext_len = strlen((char*)to);
+    if (UINT32_MAX == sealed_rsa_len) {
+        printf("ecall rsa_decrypt failed to calculate sealed data size.\n");
+        return SGX_ERROR_UNEXPECTED;
     }
-    else
-    {    
-        RSA_private_decrypt(ciphertext_len, ciphertext, plaintext, rsa, RSA_PKCS1_OAEP_PADDING);
+
+    if (cmk_blob == NULL || cmk_blob_size < sealed_rsa_len) {
+        printf("ecall rsa_decrypt cmk_blob_size is too small.\n");
+        return SGX_ERROR_INVALID_PARAMETER;
     }
 
-    return SGX_SUCCESS;
+    char* enc_key = NULL;
+    char* rsa_private_key = NULL;
+    int private_key_size;
+    int public_key_size;
+    BIO *bio = NULL;
+    RSA *rsa = NULL;
+
+    // load private key
+    do {
+        enc_key = (char*)malloc(cmk_blob_size);
+        switch(keyspec)
+        {
+            case EH_RSA_2048:
+                private_key_size = 1679;
+                public_key_size = 426;
+                break;
+            case EH_RSA_3072:
+                private_key_size = 2459;
+                public_key_size = 601;
+                break;
+        }
+        sgx_gcm_decrypt(&g_domain_key,
+                        (uint32_t*)&cmk_blob_size, (uint8_t *)enc_key,
+                        (sgx_aes_gcm_data_ex_t *)cmk_blob);
+
+        rsa_private_key = (char*)malloc(private_key_size);
+        strncpy(rsa_private_key, enc_key + public_key_size, private_key_size);
+        
+        bio = BIO_new_mem_buf(rsa_private_key, -1);
+        if (bio == NULL) {
+            printf("failed to load rsa key pem\n");
+            break;
+        }
+
+        rsa = PEM_read_bio_RSAPrivateKey(bio, NULL, NULL, NULL);
+        if (rsa == NULL) {
+            printf("failed to load rsa key\n");
+            return SGX_ERROR_INVALID_PARAMETER;
+        }
+        
+        if (plaintext_len == 0) {
+            int plaintext_len;
+            // compute plaintext length
+            unsigned char* temp_plaintext = (unsigned char*)malloc(RSA_size(rsa)); // malloc the enough space for RSA_private_decrypt
+            
+            plaintext_len = RSA_private_decrypt(ciphertext_len, ciphertext, temp_plaintext, rsa, padding_mode); // will return the plaintext length after decrypt 
+            
+            *req_plaintext_len = plaintext_len - 1; // delete '\0' for char type
```

Comment:
I do not believe this is required.   The plaintext should be always regarded as binary/uint8_t[n] but not a string/char*.

Reply:
Yes, they are same, I'll change it